cmake build capability

aerorahul opened the pull request:
This PR:
-  updates the `cmake` build capability using imported targets provided from NCEPLIBS
- makes no changes to the existing `source code`
- removes files that are placed in the library into the executable.  IO files don't belong in the post library.
- adds `PackageConfig.cmake.in` file that generates a package configuration file that can be used in external projects e.g. UFS that rely on `nceppost` library.
- adds option to build `nceppost.x` executable.
- adds option to build with/without WRF library. No wrf library code is added or attached. This is just future compatibility and is only compiled if `BUILD_WITH_WRFLIB` cmake option is turned `OFF` (`default:  OFF`). 
- adds a git submodule [CMakeModules](https://github.com/noaa-emc/CMakeModules).  This sub-module provides `FindPackage.cmake` capability to locate packages such as `NetCDF`, that CMake does not intrinsically provide search capability for.

@fossell @WenMeng-NOAA please add reviewers as appropriate

aerorahul committed:
move Makefile and Makefile_lib to deprecated state

aerorahul committed:
add .gitignore

Change-Id: Iacee1904702d6a63ed67acd3bbeffd4e10b46a97

aerorahul committed:
add CMakeModules git submodule

Change-Id: Ia0afba816595888069adb56782074b6c211bfcad

aerorahul committed:
add cmake build

Change-Id: I83715835b5810301dadc13b723cd3f4e15d21a9f

aerorahul committed:
add missing VERSION file

Change-Id: Ib5fe2b46c780e689580900ffa79ae7ea22d7ddb4

aerorahul committed:
add option to build with/without WRF library. Use WRFIO stubs if building without WRF library

aerorahul committed:
Merge branch 'develop' into feature/cmake

aerorahul committed:
missed a file that was in develop, but not in CMakeLists.txt

aerorahul committed:
make sp_4, w3nco_4, w3emc_4 PRIVATE

kgerheiser commented on sorc/ncep_post.fd/CMakeLists.txt:
```diff
@@ -0,0 +1,265 @@
+
+if(BUILD_POSTEXEC)
+  add_executable(${EXENAME} ${EXE_SRC})
+  target_link_libraries(${EXENAME} PRIVATE
```
Do you not need to link to `sp`, `w3emc`, and `w3nco` on the executable since you made them private for the lib and it would be the `d` version? Or did I misunderstand some of that.

DusanJovic-NOAA replied:
``sp_4``, ``w3emc_4``, and ``w3nco_4`` will be linked to ``ncep_post`` executable as a private dependency of ``nceppost`` target but will not be passed as a dependency to users of this librray, which is what we want so that ufs model can have direct dependency on ``_d`` versions.

fossell commented:
@aerorahul - We're completing reg tests on Hera as we speak for the merge of release/public-v1. Some of the tests are held up in queue, but slowly starting to run.  I think it might be best to allow these to finish and complete that merge, then have you update your PR to make any changes to that version.  We can start reg tests on this PR, since it is the protocol to run reg tests on PR into develop, but we may be waiting on system queue anyway. I understand that this PR in particular only adds new files and does not modify anything existing, but to bypass reg test testing I will be @WenMeng-NOAA's approval.

fossell commented on VERSION:
```diff
@@ -0,0 +1 @@
+8.1.0
```
How is this VERSION file used?  Does cmake or workflows queue off of this or...?  Each application/model implementation uses a different release version number, so I'm wondering if it makes sense to have this file in develop and if yes, what that version number should be so as not to confuse other applications.  For example, in public/release-v1 it was set to 1.0.0.   What does 8.1.0 refer to?

aerorahul replied:
Version on the develop branch is `<last-release>+dev`.
I guessed `8.1.0` from Hera (`8.0.6`)
I suppose one can add `+dev` to whatever the `last-release` is.

fossell replied:
Ok, so post library version, presumably?  But when making a branch for another release application, this file will change to the release version number (e.g. 1.0.0 for release/public-v1), which may be confusing. Is this VERSION file necessary to exist in develop? What uses this file? Or is it for informational purposes?  You can see some extra UPP release version info on the wiki: https://github.com/NOAA-EMC/EMC_post/wiki/UPP-Release

aerorahul replied:
@fossell It is the package version, whatever you call a package -- library or application.  I don't think you can have a different version of a library and an application reside in the same package, or atleast I cannot think of an example.
  
A version **must** be provided with the package.  That is just standard practice.
There are different ways to do this; I can think of 2.
- Like the one here, provide an external file and maintain it.
- Hard-wire the version number in CMakeLists.txt and update it (Unidata does this in netcdf-c)

The issue with the latter is, when issuing different releases/tags, you may rarely need to update the top-level `CMakeLists.txt`, unless you add or remove a dependency or directory, etc.  Most changes are confined to lower levels where the source code changes.  So you will have to make a change in this file, just for the sake of incrementing the version number. And this will show up as a change in the git log.

In the end, it is just a preference, how one wishes to maintain the versioning of the software.  I prefer the former.

fossell replied:
Thanks for the info.  Unfortunately the version/release procedures for UPP are not as straight forward. This was discussed in detail during UFS V1.0 release, but perhaps @WenMeng-NOAA can provide a bit of background on the impacts for implementation of operational models. I'm just not sure what we would put in VERSION for develop branch that would be accurate and consistent for all apps.

aerorahul commented:
> @aerorahul - We're completing reg tests on Hera as we speak for the merge of release/public-v1. Some of the tests are held up in queue, but slowly starting to run. I think it might be best to allow these to finish and complete that merge, then have you update your PR to make any changes to that version. We can start reg tests on this PR, since it is the protocol to run reg tests on PR into develop, but we may be waiting on system queue anyway. I understand that this PR in particular only adds new files and does not modify anything existing, but to bypass reg test testing I will be @WenMeng-NOAA's approval.

@fossell I don't see a PR to merge `release/public-v1` into `develop`

fossell commented:
> @fossell I don't see a PR to merge release/public-v1 into develop

@aerorahul - It has not been published yet. @mkavulich has started it and is patiently waiting completion of Hera reg tests.  We can go ahead and publish the PR if that is helpful?

aerorahul commented:
> > @fossell I don't see a PR to merge release/public-v1 into develop
> 
> @aerorahul - It has not been published yet. @mkavulich has started it and is patiently waiting completion of Hera reg tests. We can go ahead and publish the PR if that is helpful?

Yes. That is generally helpful so developers across the community know if there is something in the pipeline and do not end up duplicating effort.  At the very least, a draft PR is recommended.

DusanJovic-NOAA commented on cmake/PackageConfig.cmake.in:
```diff
@@ -0,0 +1,38 @@
+find_dependency(nemsio CONFIG)
+find_dependency(gfsio CONFIG)
+find_dependency(w3emc CONFIG)
+find_dependency(crtm CONFIG)
```
``nceppost::nceppost`` library target does not depend on ``gfsio``, ``nemsio``, ``sfcio``, and ``sigio``.

aerorahul replied:
resolved in 92e1d4d

aerorahul committed:
clean up nceppost library dependencies.

Change-Id: I9cbda1b897a12c329afbc17213ea8705253aa4c1

aerorahul committed:
CMakeModules was updated. Update pointer to CMakeModules submoule and in .gitmodules

Change-Id: I489d2da5b2bb7b5c28c8b06d685b665e9f86695f

aerorahul committed:
report location and version information from the library if found

Change-Id: I3b2cb47344b82c6efb1a70814c9b48b9249ba3d2

fossell commented:
@aerorahul - Can you please update your branch with develop now that release/public-v1 has been merged?  And add the version patch as well if you have not already.  Thanks!

aerorahul committed:
Merge branch 'develop' into feature/cmake
resolve conflicts.

Change-Id: I1727941175562bf62dfc55d040a37fd35fbbe8e0

aerorahul committed:
add missing files, change permissions on wrongly made executable files

Change-Id: I2ffac132e943486b6ae4d6e6b9be3f8fd1f3e507

aerorahul committed:
replace VERSION number with something outrageous because they cannot decide on an appropriate versioning.
Until then, stay outrageous.

Change-Id: I6642fa1a0ff889e49333d478b9b0e429f7a683d4

aerorahul committed:
This commit:
- names the executable consistent with the project and the library
- differentiates the executable name from the library name.
- ensures Fortran component of NetCDF is found.

Change-Id: If39f21e5bce6f12ac8053ea5a44aa6f0210c0f23

aerorahul committed:
clean-up top-level cmakelists of any downstream activity

Change-Id: Ifbd6da655b3c32ce151994082274f993fa1c5d79

aerorahul commented:
> @aerorahul - Can you please update your branch with develop now that release/public-v1 has been merged? And add the version patch as well if you have not already. Thanks!

@fossell done.

WenMeng-NOAA commented:
The regression tests were conducted on Dell, Cray and Hera. No changed results. This branch is ready for merging.
